server/lorax_server/utils/layers.py

```diff
@@ -349,6 +349,152 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
+                            v[i, :] = torch.nn.functional.embedding(
+                                input[i],
+                                data.lora_a[rank_segments.indices[i].item()][self.layer_id]
+                            )
+
+                        if self.process_group.size() > 1:
+                            v = self.collect_lora_a(v)
+
+                        add_lora_b_bgmv(
+                            proj,
+                            v,
+                            lora_b_ptr,
+                            rank_segments.indices,
+                            self.layer_id,
+                        )
+
+            if end_idx - start_idx != result.shape[1]:
+                result[:, start_idx:end_idx] += proj
+        else:
+            for adapter_index in adapter_data.meta.adapter_set:
+                if data is not None and data.has_adapter(adapter_index):
+                    adapter_mask = (adapter_data.meta.adapter_indices == adapter_index).to(input.dtype).view(-1, 1)
+                    layer_result = self.forward_lora(input, data, adapter_index, adapter_mask)
+                    result[:, start_idx:end_idx] += layer_result
+
+        return result
+
+    def forward_lora(
+        self,
+        input: torch.Tensor,
+        data: "BatchLoraWeights",
+        adapter_index: int,
+        adapter_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        lora_a = data.lora_a[adapter_index][self.layer_id, :, :]
+        lora_b = data.lora_b[adapter_index][self.layer_id, :, :]
+
+        lora_a = orient_for_rank(lora_a, lora_b.size(0))
+
+        a_out = torch.nn.functional.embedding(input, lora_a)
+        if self.process_group.size() > 1:
+            a_out = self.collect_lora_a(a_out)
+
+        result = (a_out @ lora_b) * adapter_mask
+        return result
+
+    def collect_lora_a(self, a_out: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError("Implemented in subclasses")
+
+
+class TensorParallelAdapterRowEmbedding(LoraEmbedding):
+    def __init__(self, base_layer, layer_id, layer_name, process_group):
+        super().__init__(layer_id, process_group)
+        self.base_layer = base_layer
+        self.layer_name = layer_name
+
+    @classmethod
+    def load(cls, base_layer, layer_id, layer_name, process_group):
+        return cls(base_layer, layer_id, layer_name, process_group)
+
+    def forward(self, input: torch.Tensor, adapter_data: "AdapterBatchData") -> torch.Tensor:
+        result = self.base_layer(input)
+
+        # Fused all-gather + all-reduce from S-LoRA paper: https://arxiv.org/abs/2311.03285
+        stride = result.shape[-1] // self.process_group.size()
+        start_idx = self.process_group.rank() * stride
+        end_idx = (self.process_group.rank() + 1) * stride
+
+        self.forward_layer_type(result, input, adapter_data, self.layer_name, start_idx, end_idx)
+
+        return result
+
+    # def collect_lora_a(self, a_out: torch.Tensor) -> torch.Tensor:
```

Comment:
No, it wasn't raising an error. I deliberately left it out since I didn't fully understand whether it would work. In `TensorParallelEmbedding` we're sharding the embedding weight matrix but we don't do that for linear layers so I wasn't sure the that the weights would be sharded for `TensorParallelAdapterRowEmbedding`.